backend/accounts/transaction.go

```diff
@@ -133,5 +142,122 @@ type OrderedTransactions []*TransactionData
 // The input list is modified in place and must not be used anymore after calling this function.
 func NewOrderedTransactions(txs []*TransactionData) OrderedTransactions {
 	sort.Sort(sort.Reverse(byHeight(txs)))
+
+	balance := big.NewInt(0)
+	for i := len(txs) - 1; i >= 0; i-- {
+		tx := txs[i]
+		switch tx.Type {
+		case TxTypeReceive:
+			balance.Add(balance, tx.Amount.BigInt())
+		case TxTypeSend:
+			balance.Sub(balance, tx.Amount.BigInt())
+			// Subtract fee as well.
+			if tx.Fee != nil {
+				balance.Sub(balance, tx.Fee.BigInt())
+			}
+		case TxTypeSendSelf:
+			// Subtract only fee.
+			if tx.Fee != nil {
+				balance.Sub(balance, tx.Fee.BigInt())
+			}
+		}
+		tx.Balance = coin.NewAmount(balance)
+	}
 	return txs
 }
+
+// TimeseriesEntry contains the balance of the account at the given time.
+type TimeseriesEntry struct {
+	Time  time.Time
+	Value coin.Amount
+}
+
+// MarshalJSON serializes the entry as JSON.
+func (c *TimeseriesEntry) MarshalJSON() ([]byte, error) {
+	value, err := c.Value.Int64()
+	if err != nil {
+		return nil, err
+	}
+	return json.Marshal(map[string]interface{}{
+		"time":  c.Time.Unix(),
+		"value": value,
+	})
+}
+
+// findLastIndexBeforeTime returns the index of the last transaction with a time before `target`,
+// starting at `start`.
+// Unconfirmed txs are ignored.
+// If there is no such transaction before `target`, nil is returned.
+// Note: the transactions are ordered from newest to oldest, so the returned index will be smaller
+// or equal to `start` (advancing backwards).
+func (txs OrderedTransactions) findLastIndexBeforeTime(
+	start int, target time.Time) (*int, error) {
+	currentIndex := start
+	for {
+		tx := txs[currentIndex]
+
+		if tx.Height == 0 {
+			// We reached the unconfirmed txs. We ignore those. The current index is the last tx
+			// before the target time.
+			return &currentIndex, nil
+		}
+		txTime := tx.Timestamp
+		if txTime == nil {
+			return nil, errp.WithStack(errors.ErrNotAvailable)
+		}
+		if target.Before(*txTime) {
+			if currentIndex == start {
+				// No tx before target, return nil.
+				return nil, nil
+			}
+			// We reached the first tx beyond target. Return the previous index, which is the last
+			// tx before target.
+			currentIndex++
+			return &currentIndex, nil
+		}
+		if currentIndex == 0 {
+			return &currentIndex, nil
+		}
+		currentIndex--
+	}
+}
+
+// Timeseries chunks the time between `start` and `end` into steps of `interval` duration, and
+// provides the balance of the account at each step.
+func (txs OrderedTransactions) Timeseries(
+	start, end time.Time, interval time.Duration) ([]TimeseriesEntry, error) {
+	if len(txs) == 0 {
+		return nil, nil
+	}
+	currentTime := start
+	if currentTime.IsZero() {
+		return nil, nil
+	}
+
+	result := []TimeseriesEntry{}
+	currentIndex := len(txs) - 1
+	for {
+		nextIndex, err := txs.findLastIndexBeforeTime(currentIndex, currentTime)
```

Comment:
I think it might be possible to remove `findLastIndexBeforeTime` and compute everything in one loop here, in some potential future improvements.

Reply:
I tried but it was definitely harder for me to write than this ;)